torchvision/prototype/models/depth/stereo/crestereo.py

```diff
@@ -0,0 +1,1020 @@
+import math
+from functools import partial
+from typing import Callable, Dict, Iterable, List, Optional, Tuple, Union
+
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import torchvision.models.optical_flow.raft as raft
+from torch import Tensor
+from torchvision.models._api import WeightsEnum
+from torchvision.models.optical_flow._utils import grid_sample, make_coords_grid, upsample_flow
+from torchvision.ops import Conv2dNormActivation
+
+all = (
+    "CREStereo",
+    "CREStereo_Weights",
+    "CREStereo_B_Weights",
+    "crestereo_b",
+)
+
+
+class ConvexMaskPredictor(nn.Module):
+    def __init__(
+        self,
+        *,
+        in_channels: int,
+        hidden_size: int,
+        upsample_factor: int,
+        multiplier: float = 0.25,
+    ) -> None:
+
+        super().__init__()
+        self.mask_head = nn.Sequential(
+            Conv2dNormActivation(in_channels, hidden_size, norm_layer=None, kernel_size=3),
+            nn.Conv2d(hidden_size, upsample_factor**2 * 9, 1, padding=0),
+        )
+
+        self.multiplier = multiplier
+
+    def forward(self, x: Tensor) -> Tensor:
+        x = self.mask_head(x) * self.multiplier
+        return x
+
+
+class AdaptiveGroupCorrelationLayer(nn.Module):
```

Comment:
I'm not particularly sure what the best approach of breaking it down might be. Breaking it in independent modules will add more complexity in the forward loop. What I believe might be done to reduce the size of this module would be to move `get_correlation` outside the module in a simple function.

My reasoning for keeping everything together is such that if one was to look at the paper, each individual layers code would be functionally grouped together.

Splitting this into multiple layers that you need to keep track of for each individual pyramid level in the forward pass seems prone to risks in terms of Jit and readability since now we have to dynamically elect which layer we use based on two criteria:
1.  At what pyramid level we are (since only the lowest level contains an attention module)
2. At what recursive step level we are (which basically selects the window shape)
3. Which inputs are we passing going to pass into the layers (all pyramid levels but the last one have offsets)

Reply:
One possible way I can think on top of my head now is to have:
- `IterativeCorrelation (nn.Module)`
- `AttentionOffsetCorrelation (nn.Module)`
- Then they are combined again in `AdaptiveGroupCorrelationLayer (nn.Module)`

But no strong opinion on this though, it is fine to leave as it is now.